File: Sources/AudioToText/Transcriber.swift
import AWSTranscribeStreaming
import Foundation

enum TranscriberError: Error, CustomStringConvertible {
    case missingEnvironmentVariable(String)
    case noSpeechDetected

    var description: String {
        switch self {
        case .missingEnvironmentVariable(let name):
            return "Required environment variable \(name) is not set"
        case .noSpeechDetected:
            return "No speech detected in audio"
        }
    }
}

final class Transcriber: @unchecked Sendable {
    private let client: TranscribeStreamingClient

    init() throws {
        let env = ProcessInfo.processInfo.environment
        for key in ["AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY", "AWS_REGION"] {
            guard env[key] != nil else {
                throw TranscriberError.missingEnvironmentVariable(key)
            }
        }
        let region = env["AWS_REGION"]!
        self.client = try TranscribeStreamingClient(region: region)
    }

    func transcribe(audioData: Data, sampleRateHz: Int) async throws -> String {
        let chunkSize = 16 * 1024 // 16KB

        let audioStream = AsyncThrowingStream<TranscribeStreamingClientTypes.AudioStream, Error> { continuation in
            var offset = 0
            while offset < audioData.count {
                let end = min(offset + chunkSize, audioData.count)
                let chunk = audioData.subdata(in: offset..<end)
                continuation.yield(.audioevent(TranscribeStreamingClientTypes.AudioEvent(audioChunk: chunk)))
                offset = end
            }
            continuation.finish()
        }

        let input = StartStreamTranscriptionInput(
            audioStream: audioStream,
            languageCode: .enUs,
            mediaEncoding: .pcm,
            mediaSampleRateHertz: sampleRateHz
        )

        let output = try await client.startStreamTranscription(input: input)

        var transcripts: [String] = []
        if let resultStream = output.transcriptResultStream {
            for try await event in resultStream {
                switch event {
                case .transcriptevent(let te):
                    for result in te.transcript?.results ?? [] {
                        guard !result.isPartial else { continue }
                        for alt in result.alternatives ?? [] {
                            if let text = alt.transcript, !text.isEmpty {
                                transcripts.append(text)
                            }
                        }
                    }
                case .sdkUnknown:
                    break
                }
            }
        }

        let text = transcripts.joined(separator: " ").trimmingCharacters(in: .whitespaces)
        if text.isEmpty {
            throw TranscriberError.noSpeechDetected
        }
        return text
    }

    private let chunkDurationSeconds = 30.0
    private let largeAudioThresholdSeconds = 30.0

    func transcribeChunked(audioData: Data, sampleRateHz: Int) async throws -> String {
        let durationSeconds = Double(audioData.count) / Double(sampleRateHz * 2)

        guard durationSeconds > largeAudioThresholdSeconds else {
            return try await transcribe(audioData: audioData, sampleRateHz: sampleRateHz)
        }

        let bytesPerSecond = sampleRateHz * 2
        let chunkBytes = Int(chunkDurationSeconds) * bytesPerSecond
        // Align to Int16 (2-byte) boundary
        let alignedChunkBytes = (chunkBytes / 2) * 2

        var chunks: [(index: Int, data: Data)] = []
        var offset = 0
        var index = 0
        while offset < audioData.count {
            let end = min(offset + alignedChunkBytes, audioData.count)
            chunks.append((index: index, data: audioData.subdata(in: offset..<end)))
            offset = end
            index += 1
        }

        var results: [(index: Int, transcript: String)] = []
        try await withThrowingTaskGroup(of: (Int, String).self) { group in
            for chunk in chunks {
                group.addTask {
                    do {
                        let text = try await self.transcribe(audioData: chunk.data, sampleRateHz: sampleRateHz)
                        return (chunk.index, text)
                    } catch TranscriberError.noSpeechDetected {
                        return (chunk.index, "")
                    }
                }
            }
            for try await result in group {
                results.append((index: result.0, transcript: result.1))
            }
        }

        results.sort { $0.index < $1.index }
        let combined = results.map { $0.transcript }.joined(separator: " ").trimmingCharacters(in: .whitespaces)

        if combined.isEmpty {
            throw TranscriberError.noSpeechDetected
        }
        return combined
    }
}

File: transcriber.go
package main

import (
	"context"
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/transcribestreaming"
	"github.com/aws/aws-sdk-go-v2/service/transcribestreaming/types"
)

const (
	chunkSize          = 16 * 1024 // 16KB chunks
	transcribeTimeout  = 120 * time.Second
	transcribeLanguage = "en-US"
)

type Transcriber struct {
	client *transcribestreaming.Client
}

func NewTranscriber(ctx context.Context) (*Transcriber, error) {
	for _, env := range []string{"AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY", "AWS_REGION"} {
		if os.Getenv(env) == "" {
			return nil, fmt.Errorf("required environment variable %s is not set", env)
		}
	}

	cfg, err := config.LoadDefaultConfig(ctx)
	if err != nil {
		return nil, fmt.Errorf("load AWS config: %w", err)
	}

	client := transcribestreaming.NewFromConfig(cfg)
	return &Transcriber{client: client}, nil
}

func (t *Transcriber) Transcribe(ctx context.Context, audioData []byte) (string, error) {
	ctx, cancel := context.WithTimeout(ctx, transcribeTimeout)
	defer cancel()

	resp, err := t.client.StartStreamTranscription(ctx, &transcribestreaming.StartStreamTranscriptionInput{
		LanguageCode:         types.LanguageCodeEnUs,
		MediaEncoding:        types.MediaEncodingPcm,
		MediaSampleRateHertz: int32Ptr(sampleRate),
	})
	if err != nil {
		return "", fmt.Errorf("start transcription stream: %w", err)
	}

	stream := resp.GetStream()

	go func() {
		defer stream.Close()
		for offset := 0; offset < len(audioData); offset += chunkSize {
			end := offset + chunkSize
			if end > len(audioData) {
				end = len(audioData)
			}
			err := stream.Send(ctx, &types.AudioStreamMemberAudioEvent{
				Value: types.AudioEvent{
					AudioChunk: audioData[offset:end],
				},
			})
			if err != nil {
				fmt.Println("Error sending audio chunk:", err)
				return
			}
		}
	}()

	var transcripts []string
	for event := range stream.Events() {
		switch ev := event.(type) {
		case *types.TranscriptResultStreamMemberTranscriptEvent:
			for _, result := range ev.Value.Transcript.Results {
				if result.IsPartial {
					continue
				}
				for _, alt := range result.Alternatives {
					if alt.Transcript != nil {
						transcripts = append(transcripts, *alt.Transcript)
					}
				}
			}
		}
	}

	if err := stream.Err(); err != nil {
		return "", fmt.Errorf("transcription stream error: %w", err)
	}

	text := strings.TrimSpace(strings.Join(transcripts, " "))
	if text == "" {
		return "", fmt.Errorf("no speech detected in audio")
	}

	return text, nil
}

func int32Ptr(v int) *int32 {
	i := int32(v)
	return &i
}
